---
title: "2023-05-06 TIL - Joi를 활용하여 검증하기"
categories:
  - TIL

tags:
  - [TIL]

toc: true
toc_sticky: true

author_profile: true
date: 2023-05-06
---

## Joi를 활용하여 검증하기
<br>

### Joi

사용자가 입력한 데이터를 검증하는 라이브러리이다. 여태까지 유효성 검사를 하게 되면 If문을 자주 사용해왔는데, 코드 가독성과 중복을 줄이기 위해 사용해보았다.
<br>

- #### users.validator.js

```jsx

const Joi = require("joi");

// 회원가입 검증

module.exports = {
  SignupSchema: Joi.object({
    nickname: Joi.string()
      .min(3)
      // 사용자가 지정한 패턴을 충족해야 합니다.
      .pattern(new RegExp("^[a-zA-Z0-9{3,}$]")) 
      // key에 errorcode를 넣고 value에 message를 넣어
      // 에러메세지를 커스텀할 수 있다(JSON 형태로 저장) 
      .messages({
        "string.base": "닉네임의 형식이 일치하지 않습니다.",
        "string.min": "닉네임은 최소 3자리 이상입니다.",
        "string.pattern.base": "닉네임의 형식이 일치하지 않습니다.",
      }),

    password: Joi.string().min(4).required().messages({
      "string.min": "패스워드는 최소 4자리 이상입니다.",
      "string.base": "패스워드 형식이 일치하지 않습니다.",
      "string.empty": "패스워드 형식이 일치하지 않습니다.",
    }),
    confirm: Joi.string().valid(Joi.ref("password")).required().messages({
      "any.only": "패스워드가 일치하지 않습니다.",
      "string.empty": "패스워드 형식이 일치하지 않습니다.",
    }),
  })
    .custom((value, helpers) => {
      console.log(value.nickname);
      if (value.password.toUpperCase().includes(value.nickname.toUpperCase())) {
        return helpers.error("existingNicknameInPassword");
      }
      return value;
    })
    .messages({
      existingNicknameInPassword: "패스워드 형식이 일치하지 않습니다.",
    }),
};

```
<br>

- #### 문제사항

    - custom을 활용하여 만든 함수를 SignupSchema 안에서 적용하니 value.nickname 값이 undefined 나오는 현상이 발생했다.
<br>

- #### 해결
    - value는 검증을 성공한 경우 입력된 데이터이다.
   SignupSchema를 통과하지 않으면 value가 생성되지 않는데, custom 메서드를 SignupSchema안에 포함시켰기 때문에 발생한 문제였다. 위와 같이 밖으로 빼서 사용하니 잘 작동하였다.
<br>

- #### users.controller.js


```jsx

// 회원가입 controller 함수

createSignup = async (req, res, next) => {
    try {

        // Joi의 validateAsync메서드를 활용하여 검증한 후 검증을
        // 통과한 nickname과 password 객체 구조 분해 할당을 통해 저장
      const { nickname, password } = await SignupSchema.validateAsync(
        req.body
      ).catch((error) => {
        throw new Error(`412/${error}`);
      });

      await this.userService.createSignup(nickname, password);
      res.status(201).json({ message: "회원 가입에 성공하였습니다." });
    } catch (error) {
      console.error(error);
      return res
        .status(400)
        .json({ errorMessage: "요청한 데이터 형식이 올바르지 않습니다." });
    }
}
```
<br>

#### Reference
<br>

<https://joi.dev/api/?v=17.9.1>
<https://velog.io/@mero/joi-messages-%EA%B8%B0%EB%8A%A5-%ED%99%9C%EC%9A%A9>